rbac/migration_tool/sharedSystemRolesReplicatedRoleBindings.py

```diff
@@ -147,74 +168,184 @@ def v1_role_to_v2_bindings(
                         resource_id = str(ungrouped_ws.id)
                     else:
                         continue
-                add_element(perm_groupings, V2boundresource(resource_type, resource_id), v2_perm, collection=set)
+                add_element(perm_groupings, V2boundresource(resource_type, resource_id), permission, collection=set)
         if default:
             add_element(
                 perm_groupings,
                 default_resource,
-                v2_perm,
+                permission,
                 collection=set,
             )
 
     # Project permission sets to roles per set of resources
-    return permission_groupings_to_v2_role_bindings(perm_groupings, v1_role, role_bindings)
+    return permission_groupings_to_v2_role_bindings(
+        perm_groupings,
+        v1_role,
+        existing_mappings=existing_role_bindings,
+        existing_v2_roles=existing_v2_roles,
+    )
+
+
+def _target_role_uuid_for(
+    existing_binding: Optional[V2rolebinding],
+    latest_roles: Iterable[RoleV2],
+) -> uuid.UUID:
+    """Return the UUID to use for a new V2 role given an existing V2rolebinding and the set of V2 roles already used."""
+    if existing_binding is not None:
+        existing_uuid = existing_binding.role.id
+        uuid_already_used = any(str(r.uuid) == existing_uuid for r in latest_roles)
+
+        if not uuid_already_used:
+            return uuid.UUID(existing_uuid)
+
+    # Return a new UUID, since we can't use the existing one.
+    return uuid.uuid7()
+
+
+def _v2_custom_role_from_v1(v1_role: Role, target_uuid: uuid.UUID):
+    # This guarantees that we do not accidentally move V2 roles between different V1 sources. If the UUID
+    # already exists with a different V1 source, then this will attempt to create a new V2 role with an
+    # existing UUID, which will fail due to the conflict.
+    #
+    # If we do not have an existing UUID to reuse, then is guaranteed to create a new role because we've just
+    # generated a new random UUID.
+    new_role, _ = CustomRoleV2.objects.update_or_create(
+        uuid=target_uuid,
+        tenant=v1_role.tenant,
+        v1_source=v1_role,
+        defaults=dict(
+            name=str(target_uuid),
```

Comment:
Those names will be visible in UI, we might need something what is readable.
In DDR we discussed something like:

"<name of custom role> (1)"
"<name of custom role> (2)"
"<name of custom role> (3)"
....

I am not sure what could other options. 


cc @astrozzc